Compatibility fixes for solc-js TypeScript migration

cameel opened the pull request:
This PR adjusts the command we run from solc-js and adds `npm run build` required to compile the TypeScript code. It should be merged after https://github.com/ethereum/solc-js/pull/566.

cameel committed:
bytecode compare: Stop depending on solc-js internal structure and install it via npm from a local dir instead

cameel commented:
I found two more places where we do not run `solcjs` but we do `npm install` on solc-js. I think they now need `npm run build` too. PR updated.

cameel commented on scripts/wasm-rebuild/docker-scripts/rebuild_tags.sh:
```diff
@@ -228,6 +228,7 @@ mkdir -p "${OUTPUTDIR}"/bin
 echo "Prepare solc-js."
 cd /root/solc-js
 npm install >/dev/null 2>&1
+npm run build >/dev/null 2>&1
```
Not sure about this one and these scripts are not actively tested but I'm adding it just in case.

ekpyron commented:
Tests are still failing.
Also: does the release checklist need adjustment or is it fine?

cameel committed:
Changelog entry for the TypeScript migration

cameel commented:
> Also: does the release checklist need adjustment or is it fine?

Right, it does. You cannot run `verifyVersion.js` directly. You need `npx ts-node verifyVersion.ts` instead.
But I think we should replace it with `npm run updateBinary`. It does this but also downloads the binary. I think we should have changed it to that even earlier, back when we made it not download a new binary automatically.

cameel committed:
Release checklist: switch to a command that automatically downloads the binary before running verifyVersion

cameel committed:
Adjust workflow with solc-js checkout now that it's on TypeScrpt

cameel commented:
- Checklist updated.
- Added changelog entry for TypeScript migration.
- Bytecode comparison was failing because the JS snippet was importing specific .js files directly instead of doing it via the public interface. I changed it completely so that now the package is installed via npm from a local dir and we can just import `solc`.
- solc-js and Colony external tests were failing because now `soljson.js` is actually under `dist/`. `npm run build` copies it there from the old location but only if we put it in that old location before running the command. That was not the case.
    - It's only Colony test that failed because it's the only one that's still on solc-js rather than native binary. I think we should keep one working like that just to have this code path somewhat tested.

ekpyron commented:
Still something off in the tests apparently...

cameel committed:
fixup! bytecode compare: Stop depending on solc-js internal structure and install it via npm from a local dir instead

cameel commented:
Pushing a fix for bytecode comparison (turns out I still need to run `npm install` and `npm run build` inside the directory I install a local package from).

Still working on the other fail.

cameel commented:
The test failure in solc-js external test is actually a bug in solc-js itself. The failing test is buggy but gets skipped due to missing Z3. It does not get skipped here :) We need to merge https://github.com/ethereum/solc-js/pull/592 to fix it.

cameel commented:
The problem with Colony is harder. I think we may have actually broken backwards-compatibility with the Typescript change.
It fails with `TypeError: soljson.cwrap is not a function` in Trufle's [`Local.load()`](https://github.com/trufflesuite/truffle/blob/v5.4.30/packages/compile-solidity/src/compilerSupplier/loadingStrategies/Local.ts#L7-L25).

Truffle does `soljson = originalRequire(compilerPath)`, where `compilerPath` is actually a path to the directory containing solc-js checkout (because that's what I'm giving it in the test). In JS that's fine and `soljson.compile` function is present. Now, however, it's under `soljson.default.compile` due to `default export` in our `wrapper.js` and `soljson.compile` is `undefined`.

From what I'm reading, not having `default` in there was either some kind of hack or it's something that changed by design in TypeScript. In any case, we probably need to change how we export stuff because otherwise JS tools will be able to import it without changes on their side.

cameel commented:
@stephensli What do you think about https://github.com/ethereum/solidity/pull/12583#issuecomment-1021430071? Is that actually a breaking change or is there some easy workaround?

axic commented:
So maybe we do need to do those manual function exports, so they are named.

stephensli commented:
https://github.com/ethereum/solc-js/pull/593

This is the way to get around, it should now function as expected

ekpyron commented:
[Force-pushed to retrigger tests after merging https://github.com/ethereum/solc-js/pull/593]
